third_party/amd/lib/TritonAMDGPUTransforms/StreamPipeline.cpp

```diff
@@ -549,6 +549,41 @@ void LoopPipeliner::emitPrologue() {
 }
 
 void LoopPipeliner::emitEpilogue(DenseMap<Value, Value> &newResults) {
+  // After pipelining, some of the depArgs have beem mapped to new args.
+  // We need to remap these.
+  //
+  // For example, if we have
+  //
+  //   ptr = ...
+  //   ret = scf.for iter_args(a_ptr=ptr)
+  //     a = load(a_ptr)
+  //     c += dot(a, ...)
+  //     a_ptr_new = a_ptr + N
+  //     scf.yield %a_ptr_new
+  //
+  // then the ptr arg should be mapped to a new arg in the for loop.
+  //
+  //   ptr = ...
+  //   load_pre = load(ptr)
+  //   ptr_new = ptr + N
+  //   ret = scf.for iter_args(a_ptr=ptr, ld=load_pre, A_ptr=ptr_new)
+  //     a_next = load(a_ptr)
+  //     c += dot(ld, ...)
+  //     A_ptr_new = A_ptr + N
+  //     scf.yield a_ptr, a_next, A_ptr_new
+  //
+  // After this, if there are downstream users of a_ptr, they should reference
+  // ret#2 instead of ret#0
+  for (const auto &origArg : llvm::enumerate(forOp.getRegionIterArgs())) {
+    if (depArgs.contains(origArg.value())) {
+      auto oldIdx = origArg.index();
+      auto newIdx = depArgsIdx[origArg.value()];
+      auto oldResult = forOp->getResult(oldIdx);
+      auto newResult = pplForOp->getResult(newIdx);
+      newResults[oldResult] = newResult;
```

Comment:
newResults is also updated later at end of this function. I wonder what is the relation of these two. Are they handling different cases?

Reply:
Yes, different cases.

Later in this function, it updates the results for anything that is written in the epilogue.  So if, for example, you have

```
// prologue
%1 = scf.for ... {
// ... for loop
}
%dot = tl.dot(... %1#2)
```

then you need to update some downstream users to use %dot instead of using some return value from %1.

Meanwhile, this PR is updating downstream users of variables returned from the forOp, which is a different case.